DocsPortingTool fails to convert generic targets of `<see cref` correctly

The following line: https://github.com/dotnet/dotnet-api-docs/blob/96b853d5e314fc00f8fa2acad402ddb3a00b76a7/xml/System.IO.Pipelines/PipeOptions.xml#L160

```
<summary>Gets the <see cref="T:System.Buffers.MemoryPool`1" /> object used for buffer management.</summary>
```

When ported did not change the syntax of the generic type.  CSC then complains:

```
error CS1584: XML comment has syntactically incorrect cref attribute 'System.Buffers.MemoryPool`1' 
```

To fix it I had to manually change the generic parameter syntax to `{T}`.  Hope I got it right?

The tool should be able to do this for me.

Command:
```
-Direction ToTripleSlash -CsProj c:\src\dotnet\runtime\src\libraries\System.IO.Pipelines\src\System.IO.Pipelines.csproj -docs c:\src\dotnet\dotnet-api-docs\xml -IncludedAssemblies System.IO.Pipelines -IncludedNamespaces System.IO
```

Using dotnet/runtime add68df26af8c0dab2afb13da1f0ca033753fd0b

There were a mess of these issues in this project.  I ended up find->replacing `` `1`` with `{T}` to deal with them.  I checked and the resultant XML was correct.  Not sure if that's a great strategy, ideally the *actual* type parameter name could be used.

--- a/BackportInstructions.md
+++ b/BackportInstructions.md
@@ -130,7 +130,7 @@ For reference, here are a couple of assemblies that have been addressed:
 
 10. After backporting, you will have to fix some things the tool can't fix:
 
-    a. MS Docs supports using `<see cref="DocId" />`elements that point to method overloads, but triple slash comments don’t support this yet (see [existing csharplang issue](https://github.com/dotnet/csharplang/issues/320)). When the tool ports these, it has no way of knowing if a cref is referencing a method overload, so your runtime project will fail with **CS0419**: "`Ambiguous reference in cref attribute`". Fix it by adding the prefix `O:` to the cref DocId, to indicate an overload.
+    a. MS Docs supports using `<see cref="DocId" />`elements that point to method overloads (no parenthesis), but triple slash comments don’t support this yet (see [existing csharplang issue](https://github.com/dotnet/csharplang/issues/320)). When the tool ports these, it has no way of knowing if a cref is referencing a method overload, so your project will fail with **CS0419**: "`Ambiguous reference in cref attribute`". Fix it by adding the prefix `O:` to the cref DocId, to indicate an overload.
 
     b. You may encounter `<param name="parameterName" />` elements that point to parameters in the signature that do not exist, causing error **CS1572**: "`XML comment has a param tag but there is no parameter by that name`". In some cases, you will notice that the next parameter will have a very similar or identical description. This means that a parameter had that name in a particular version, and was then renamed in a newer version. To fix this, remove the `<param>` that does not exist anymore (if you see duplicate descriptions), or rename the param to the correct name (if you don't find a param with a duplicate description).
 
@@ -138,7 +138,7 @@ For reference, here are a couple of assemblies that have been addressed:
 
     d. Some `<see cref="DocId" />` elements may point to APIs living in unreferenced assemblies. We don't want to add them as a dependency of the current project just to resolve the documentation problem, so instead of using `<see cref="DocId" />`, use `<see langword="DocId" />`.
 
-    e. Some markdown hyperlinks with the format `[Friendly text](URL)` may not be resolved by the tool (although most cases should be converted). If you find these in plain xml, convert them to `<a href="URL">Friendly text</a>`.
+    e. Some markdown hyperlinks with the format `[Friendly text](URL)` may not be detected by the tool to convert them to xml (although most cases should be converted). If you find these in plain xml, convert them to `<a href="URL">Friendly text</a>`.
 
     f. Some remarks have xrefs with normal characters converted to URL entities. The tool can convert URL entities back to normal characters. For example:
 
@@ -154,6 +154,8 @@ For reference, here are a couple of assemblies that have been addressed:
     h. The tool is capable of converting most markdown to xml when needed, but it may miss a few difficult cases. Make sure to find any unexpected markdown in xml sections (not wrapped by `<format>` and `CDATA`) and convert them to the appropriate xml tag. The elements that the tool will never convert to xml, and will preserve as markdown, are: `[!INCLUDE]`, `[!NOTE]`, `[!IMPORTANT]`, `[!TIP]`.
 
     i. The tool should be able to preserve double slash comments if they were placed on top of a public API, and will add the new triple slash comments separately. Make sure to verify that no important information is lost, especially the kind that is directed to developers/maintainers.
+
+    j. Docs repos indicate generic type parameters using `` `1 ``, `` `2 ``, etc., but in triple slash, they are represented with `{T}`, `{T, U}`, etc. The tool currently just converts them all to `{T}`, but it may not be the correct type parameter name or number of parameters, so make sure it's correct.
 
   > Examples: Refer to the following PRs:
   >

--- a/Libraries/RoslynTripleSlash/TripleSlashSyntaxRewriter.cs
+++ b/Libraries/RoslynTripleSlash/TripleSlashSyntaxRewriter.cs
@@ -100,14 +100,12 @@ namespace Libraries.RoslynTripleSlash
 
         private static readonly string[] MarkdownHeaders = new[] { "[!NOTE]", "[!IMPORTANT]", "[!TIP]" };
 
-        private static readonly string ValidRegexChars = @"A-Za-z0-9\-\._~:\/#\[\]@!\$&'\(\)\*\+,;%`";
+        private static readonly string ValidRegexChars = @"A-Za-z0-9\-\._~:\/#\[\]\{\}@!\$&'\(\)\*\+,;`";
         private static readonly string ValidExtraChars = @"\?=";
 
-        private static readonly string RegexMarkdownXrefPattern = @"(?<xref><xref\:(?<DocId>[" + ValidRegexChars + @"]+)(?<extraVars>\?[" + ValidRegexChars + @"]+=[" + ValidRegexChars + @"]+)?>)";
-        private static readonly string RegexXmlSeeCrefReplacement = "<see cref=\"${DocId}\" />";
-
-        private static readonly string RegexMarkdownXrefOverloadPattern = @"(?<xref><xref\:(?<DocId>[" + ValidRegexChars + @"]+)%2[aA](?<extraVars>\?[" + ValidRegexChars + @"]+=[" + ValidRegexChars + @"]+)?>)";
-        private static readonly string RegexXmlSeeCrefOverloadReplacement = "<see cref=\"O:${DocId}\" />";
+        private static readonly string RegexDocIdPattern = @"(?<prefix>[A-Za-z]{1}:)?(?<docId>[" + ValidRegexChars + @"]+)(?<overload>%2[aA])?(?<extraVars>\?[" + ValidRegexChars + @"]+=[" + ValidRegexChars + @"]+)?";
+        private static readonly string RegexXmlCrefPattern = "cref=\"" + RegexDocIdPattern + "\"";
+        private static readonly string RegexMarkdownXrefPattern = @"(?<xref><xref:" + RegexDocIdPattern + ">)";
 
         private static readonly string RegexMarkdownBoldPattern = @"\*\*(?<content>[A-Za-z0-9\-\._~:\/#\[\]@!\$&'\(\)\+,;%` ]+)\*\*";
         private static readonly string RegexXmlBoldReplacement = @"<b>${content}</b>";
@@ -597,7 +595,8 @@ namespace Libraries.RoslynTripleSlash
         {
             if (!text.IsDocsEmpty())
             {
-                TypeCrefSyntax crefSyntax = SyntaxFactory.TypeCref(SyntaxFactory.ParseTypeName(RemoveDocIdPrefixes(cref)));
+                cref = RemoveCrefPrefix(cref);
+                TypeCrefSyntax crefSyntax = SyntaxFactory.TypeCref(SyntaxFactory.ParseTypeName(cref));
                 XmlTextSyntax contents = GetTextAsCommentedTokens(text, leadingWhitespace);
                 XmlElementSyntax element = SyntaxFactory.XmlExceptionElement(crefSyntax, contents);
                 return GetXmlTrivia(element, leadingWhitespace);
@@ -622,7 +621,7 @@ namespace Libraries.RoslynTripleSlash
 
         private static SyntaxTriviaList GetSeeAlso(string cref, SyntaxTriviaList leadingWhitespace)
         {
-            cref = ReplacePrimitiveTypes(cref);
+            cref = RemoveCrefPrefix(cref);
             TypeCrefSyntax crefSyntax = SyntaxFactory.TypeCref(SyntaxFactory.ParseTypeName(cref));
             XmlEmptyElementSyntax element = SyntaxFactory.XmlSeeAlsoElement(crefSyntax);
             return GetXmlTrivia(element, leadingWhitespace);
@@ -644,7 +643,7 @@ namespace Libraries.RoslynTripleSlash
 
         private static SyntaxTriviaList GetAltMember(string cref, SyntaxTriviaList leadingWhitespace)
         {
-            cref = ReplacePrimitiveTypes(cref);
+            cref = RemoveCrefPrefix(cref);
             XmlAttributeSyntax attribute = SyntaxFactory.XmlTextAttribute("cref", cref);
             XmlEmptyElementSyntax emptyElement = SyntaxFactory.XmlEmptyElement(SyntaxFactory.XmlName(SyntaxFactory.Identifier("altmember")), new SyntaxList<XmlAttributeSyntax>(attribute));
             return GetXmlTrivia(emptyElement, leadingWhitespace);
@@ -691,7 +690,7 @@ namespace Libraries.RoslynTripleSlash
 
         private static XmlTextSyntax GetTextAsCommentedTokens(string text, SyntaxTriviaList leadingWhitespace, bool wrapWithNewLines = false)
         {
-            text = ReplacePrimitiveTypes(text);
+            text = CleanCrefs(text);
 
             // collapse newlines to a single one
             string whitespace = Regex.Replace(leadingWhitespace.ToFullString(), @"(\r?\n)+", "");
@@ -905,11 +904,7 @@ namespace Libraries.RoslynTripleSlash
 
         private static string ReplaceMarkdownWithXmlElements(string text, List<DocsParam> docsParams, List<DocsTypeParam> docsTypeParams)
         {
-            // see crefs with overloads
-            text = Regex.Replace(text, RegexMarkdownXrefOverloadPattern, RegexXmlSeeCrefOverloadReplacement);
-
-            // see crefs
-            text = Regex.Replace(text, RegexMarkdownXrefPattern, RegexXmlSeeCrefReplacement);
+            text = CleanXrefs(text);
 
             // commonly used url entities
             text = Regex.Replace(text, @"%23", "#");
@@ -947,30 +942,58 @@ namespace Libraries.RoslynTripleSlash
                 }
             }
 
-            text = ReplacePrimitiveTypes(text);
-
             return text;
         }
 
-        private static string RemoveDocIdPrefixes(string text)
+        // Removes the one letter prefix and the following colon, if found, from a cref.
+        private static string RemoveCrefPrefix(string cref)
         {
-            if (text.Length > 2 && text[1] == ':')
+            if (cref.Length > 2 && cref[1] == ':')
             {
-                return text[2..];
+                return cref[2..];
             }
-
-            text = Regex.Replace(text, @"cref=""[a-zA-Z]{1}\:", "cref=\"");
-
-            return text;
+            return cref;
         }
 
-        private static string ReplacePrimitiveTypes(string text)
+        private static string ReplacePrimitives(string text)
         {
-            text = RemoveDocIdPrefixes(text);
             foreach ((string key, string value) in PrimitiveTypes)
             {
                 text = Regex.Replace(text, key, value);
             }
+            return text;
+        }
+
+        private static string ReplaceDocId(Match m)
+        {
+            string docId = m.Groups["docId"].Value;
+            string overload = string.IsNullOrWhiteSpace(m.Groups["overload"].Value) ? "" : "O:";
+            docId = ReplacePrimitives(docId);
+            docId = Regex.Replace(docId, @"`\d", "{T}");
+            return overload + docId;
+        }
+
+        private static string CrefEvaluator(Match m)
+        {
+            string docId = ReplaceDocId(m);
+            return "cref=\"" + docId + "\"";
+        }
+
+        private static string CleanCrefs(string text)
+        {
+            text = Regex.Replace(text, RegexXmlCrefPattern, CrefEvaluator);
+            return text;
+        }
+
+        private static string XrefEvaluator(Match m)
+        {
+            string docId = ReplaceDocId(m);
+            return "<see cref=\"" + docId + "\" />";
+        }
+
+        private static string CleanXrefs(string text)
+        {
+            text = Regex.Replace(text, RegexMarkdownXrefPattern, XrefEvaluator);
             return text;
         }
 

--- a/Tests/PortToTripleSlash/TestData/Basic/MyType.xml
+++ b/Tests/PortToTripleSlash/TestData/Basic/MyType.xml
@@ -22,6 +22,8 @@ Multiple lines.
 
 This text is not a note. It has a <xref:MyNamespace.MyType> that should be xml and outside **the cdata**.
 
+Long xrefs one after the other: <xref:System.IO.Pipelines.PipeWriter.FlushAsync(System.Threading.CancellationToken)> or <xref:System.IO.Pipelines.PipeWriter.WriteAsync(System.ReadOnlyMemory{System.Byte},System.Threading.CancellationToken)> should both be converted to crefs.
+
       ]]></format>
     </remarks>
   </Docs>

--- a/Tests/PortToTripleSlash/TestData/Basic/SourceExpected.cs
+++ b/Tests/PortToTripleSlash/TestData/Basic/SourceExpected.cs
@@ -25,7 +25,8 @@ namespace MyNamespace
     /// > [!NOTE]
     /// > This note should prevent converting markdown to xml. It has a <xref:MyNamespace.MyEnum>.
     /// ]]></format>
-    /// This text is not a note. It has a <see cref="MyNamespace.MyType" /> that should be xml and outside <b>the cdata</b>.</remarks>
+    /// This text is not a note. It has a <see cref="MyNamespace.MyType" /> that should be xml and outside <b>the cdata</b>.
+    /// Long xrefs one after the other: <see cref="System.IO.Pipelines.PipeWriter.FlushAsync(System.Threading.CancellationToken)" /> or <see cref="System.IO.Pipelines.PipeWriter.WriteAsync(System.ReadOnlyMemory{byte},System.Threading.CancellationToken)" /> should both be converted to crefs.</remarks>
     // Original MyType class comments with information for maintainers, must stay.
     public class MyType
     {
@@ -139,8 +140,8 @@ namespace MyNamespace
         /// <param name="sender">This is the sender parameter.</param>
         /// <param name="e">This is the e parameter.</param>
         /// <typeparam name="T">This is the MyDelegate typeparam T.</typeparam>
-        /// <remarks>These are the <see cref="MyNamespace.MyType.MyDelegate`1" /> remarks. There is a code example, which should be moved to its own examples section:</remarks>
-        /// <example>Here is some text in the examples section. There is an <see cref="MyNamespace.MyType.MyDelegate`1" /> that should be converted to xml.
+        /// <remarks>These are the <see cref="MyNamespace.MyType.MyDelegate{T}" /> remarks. There is a code example, which should be moved to its own examples section:</remarks>
+        /// <example>Here is some text in the examples section. There is an <see cref="MyNamespace.MyType.MyDelegate{T}" /> that should be converted to xml.
         /// The snippet links below should be inserted in markdown.
         /// <format type="text/markdown"><![CDATA[
         /// [!code-csharp[MyExample#1](~/samples/snippets/example.cs)]

